File: impl/typescript/src/registry/validation.ts
/**
 * AgentURI validation rules for registration completeness.
 *
 * Spec reference: 003 data-model.md, Validation Rules V-REG-01..V-REG-12
 * Spec reference: 003 spec.md
 *   - FR-R02: Three mandatory neuron-topic services (stdIn, stdOut, stdErr).
 *   - FR-R03: Mandatory neuron-p2p-exchange with peerID, protocol, topicRef.
 *   - FR-R08: Registration completeness.
 *   - FR-R13, FR-R14: Optional DID service constraints.
 *
 * This module implements the locally-checkable validation rules (V-REG-01..07,
 * V-REG-11, V-REG-12). On-chain checks (V-REG-08..10) are deferred to
 * registration time and are documented but not implemented here.
 */

import type { NeuronPublicKey } from '../keylib/public-key.js';
import type {
  AgentURI,
  NeuronTopicServiceEntry,
  NeuronP2PExchangeEntry,
  DIDServiceEntry,
} from './types.js';

// ---------------------------------------------------------------------------
// ValidationError
// ---------------------------------------------------------------------------

/**
 * A single validation failure with the rule identifier and message.
 *
 * Each rule maps to a V-REG-NN identifier from 003 data-model.md.
 */
export interface ValidationError {
  /** Validation rule identifier (e.g., "V-REG-01"). */
  readonly rule: string;

  /** Human-readable description of the validation failure. */
  readonly message: string;
}

/**
 * Result of AgentURI validation.
 *
 * `valid` is true only when `errors` is empty.
 */
export interface ValidationResult {
  readonly valid: boolean;
  readonly errors: ReadonlyArray<ValidationError>;
}

// ---------------------------------------------------------------------------
// Standard channel roles
// ---------------------------------------------------------------------------

/** The three mandatory standard channel roles. FR-R02, FR-R08 */
const STANDARD_CHANNELS: ReadonlyArray<string> = ['stdIn', 'stdOut', 'stdErr'];

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Validate an AgentURI for registration completeness.
 *
 * Checks all locally-verifiable validation rules (V-REG-01..07, V-REG-11, V-REG-12).
 * On-chain checks (V-REG-08 proof-of-control, V-REG-09 duplicate, V-REG-10 ownership)
 * are deferred to registration time.
 *
 * FR-R08: A complete AgentURI MUST contain exactly three neuron-topic services
 * (stdIn, stdOut, stdErr), at least one neuron-p2p-exchange, and at most one DID.
 *
 * @param agentURI - The AgentURI to validate
 * @param childPublicKey - The Child's NeuronPublicKey (for V-REG-06, V-REG-12)
 * @returns ValidationResult with `valid` flag and array of ValidationError
 */
export function validateRegistrationCompleteness(
  agentURI: AgentURI,
  childPublicKey: NeuronPublicKey,
): ValidationResult {
  const errors: ValidationError[] = [];

  // Classify services by type
  const topicServices: NeuronTopicServiceEntry[] = [];
  const p2pServices: NeuronP2PExchangeEntry[] = [];
  const didServices: DIDServiceEntry[] = [];

  for (const svc of agentURI.services) {
    switch (svc.type) {
      case 'neuron-topic':
        topicServices.push(svc);
        break;
      case 'neuron-p2p-exchange':
        p2pServices.push(svc);
        break;
      case 'DID':
        didServices.push(svc);
        break;
      default:
        // Unknown service types are ignored (forward-compatible)
        break;
    }
  }

  // V-REG-01: Exactly 3 neuron-topic services with channels stdIn, stdOut, stdErr
  checkVREG01(topicServices, errors);

  // V-REG-02: At least 1 neuron-p2p-exchange service
  checkVREG02(p2pServices, errors);

  // V-REG-03: Each neuron-topic has all required fields
  checkVREG03(topicServices, errors);

  // V-REG-04: Each neuron-p2p-exchange has all required fields
  checkVREG04(p2pServices, errors);

  // V-REG-05: Each p2p topicRef matches an existing neuron-topic service name
  checkVREG05(topicServices, p2pServices, errors);

  // V-REG-06: DID service endpoint is did:key from childPublicKey
  checkVREG06(didServices, childPublicKey, errors);

  // V-REG-07: At most one DID service
  checkVREG07(didServices, errors);

  // V-REG-11: Standard channels appear exactly once each
  checkVREG11(topicServices, errors);

  // V-REG-12: P2P peerID matches Child's PeerID
  checkVREG12(p2pServices, childPublicKey, errors);

  return {
    valid: errors.length === 0,
    errors,
  };
}

// ---------------------------------------------------------------------------
// Individual rule checks
// ---------------------------------------------------------------------------

/**
 * V-REG-01: AgentURI contains exactly 3 neuron-topic services (stdIn, stdOut, stdErr).
 *
 * FR-R08, FR-R02: Registration completeness requires three topic services.
 */
function checkVREG01(
  topics: readonly NeuronTopicServiceEntry[],
  errors: ValidationError[],
): void {
  if (topics.length !== 3) {
    errors.push({
      rule: 'V-REG-01',
      message: `AgentURI must contain exactly 3 neuron-topic services, found ${topics.length.toString()}`,
    });
    return;
  }

  const channels = new Set(topics.map(t => t.channel));
  for (const expected of STANDARD_CHANNELS) {
    if (!channels.has(expected)) {
      errors.push({
        rule: 'V-REG-01',
        message: `AgentURI is missing neuron-topic service for channel "${expected}"`,
      });
    }
  }
}

/**
 * V-REG-02: AgentURI contains at least 1 neuron-p2p-exchange service.
 *
 * FR-R08, FR-R03: Registration completeness requires at least one P2P service.
 */
function checkVREG02(
  p2p: readonly NeuronP2PExchangeEntry[],
  errors: ValidationError[],
): void {
  if (p2p.length < 1) {
    errors.push({
      rule: 'V-REG-02',
      message: 'AgentURI must contain at least 1 neuron-p2p-exchange service, found 0',
    });
  }
}

/**
 * V-REG-03: Each neuron-topic has all MUST fields.
 *
 * FR-R02, 004 FR-T14: type, name, version, channel, transport, anchor, config.
 */
function checkVREG03(
  topics: readonly NeuronTopicServiceEntry[],
  errors: ValidationError[],
): void {
  const requiredFields: ReadonlyArray<keyof NeuronTopicServiceEntry> = [
    'type', 'name', 'version', 'channel', 'transport', 'anchor', 'config',
  ];

  for (const svc of topics) {
    for (const field of requiredFields) {
      const value = svc[field];
      if (value === undefined || value === null) {
        errors.push({
          rule: 'V-REG-03',
          message: `neuron-topic service "${svc.name || '(unnamed)'}" is missing required field "${field}"`,
        });
      } else if (typeof value === 'string' && value.length === 0) {
        errors.push({
          rule: 'V-REG-03',
          message: `neuron-topic service "${svc.name || '(unnamed)'}" has empty required field "${field}"`,
        });
      }
    }
  }
}

/**
 * V-REG-04: Each neuron-p2p-exchange has all MUST fields.
 *
 * FR-R03, 004 FR-T17: type, name, version, peerID, protocol, topicRef.
 */
function checkVREG04(
  p2p: readonly NeuronP2PExchangeEntry[],
  errors: ValidationError[],
): void {
  const requiredFields: ReadonlyArray<keyof NeuronP2PExchangeEntry> = [
    'type', 'name', 'version', 'peerID', 'protocol', 'topicRef',
  ];

  for (const svc of p2p) {
    for (const field of requiredFields) {
      const value = svc[field];
      if (value === undefined || value === null) {
        errors.push({
          rule: 'V-REG-04',
          message: `neuron-p2p-exchange service "${svc.name || '(unnamed)'}" is missing required field "${field}"`,
        });
      } else if (typeof value === 'string' && value.length === 0) {
        errors.push({
          rule: 'V-REG-04',
          message: `neuron-p2p-exchange service "${svc.name || '(unnamed)'}" has empty required field "${field}"`,
        });
      }
    }
  }
}

/**
 * V-REG-05: topicRef in neuron-p2p-exchange matches a neuron-topic service name.
 *
 * 004 FR-T18: Cross-reference validation.
 */
function checkVREG05(
  topics: readonly NeuronTopicServiceEntry[],
  p2p: readonly NeuronP2PExchangeEntry[],
  errors: ValidationError[],
): void {
  const topicNames = new Set(topics.map(t => t.name));

  for (const svc of p2p) {
    if (svc.topicRef && !topicNames.has(svc.topicRef)) {
      errors.push({
        rule: 'V-REG-05',
        message:
          `neuron-p2p-exchange "${svc.name}" has topicRef "${svc.topicRef}" ` +
          `that does not match any neuron-topic service name. ` +
          `Available: [${Array.from(topicNames).join(', ')}]`,
      });
    }
  }
}

/**
 * V-REG-06: DID service endpoint is did:key from childPublicKey.
 *
 * FR-R14: When present, the DID value MUST be derived from the Child's
 * NeuronPublicKey (secp256k1, multicodec 0xe7, base58btc prefix zQ3s).
 */
function checkVREG06(
  dids: readonly DIDServiceEntry[],
  childPublicKey: NeuronPublicKey,
  errors: ValidationError[],
): void {
  if (dids.length === 0) {
    return; // DID service is optional
  }

  const expectedDID = childPublicKey.didKey().toString();

  for (const svc of dids) {
    if (svc.endpoint !== expectedDID) {
      errors.push({
        rule: 'V-REG-06',
        message:
          `DID service endpoint "${svc.endpoint}" does not match Child's did:key. ` +
          `Expected: "${expectedDID}"`,
      });
    }
  }
}

/**
 * V-REG-07: At most one DID service per registration.
 *
 * FR-R14: At most one DID service is permitted per registration.
 */
function checkVREG07(
  dids: readonly DIDServiceEntry[],
  errors: ValidationError[],
): void {
  if (dids.length > 1) {
    errors.push({
      rule: 'V-REG-07',
      message: `AgentURI must contain at most 1 DID service, found ${dids.length.toString()}`,
    });
  }
}

/**
 * V-REG-11: Standard channel names each appear exactly once in neuron-topic services.
 *
 * FR-R02, FR-R08: Duplicate standard channels are invalid.
 */
function checkVREG11(
  topics: readonly NeuronTopicServiceEntry[],
  errors: ValidationError[],
): void {
  const channelCounts = new Map<string, number>();

  for (const svc of topics) {
    if (STANDARD_CHANNELS.includes(svc.channel)) {
      channelCounts.set(svc.channel, (channelCounts.get(svc.channel) ?? 0) + 1);
    }
  }

  for (const [channel, count] of channelCounts) {
    if (count > 1) {
      errors.push({
        rule: 'V-REG-11',
        message: `Standard channel "${channel}" appears ${count.toString()} times, expected exactly once`,
      });
    }
  }
}

/**
 * V-REG-12: neuron-p2p-exchange peerID matches Child's PeerID.
 *
 * FR-R03: The peerID field MUST be the Child's libp2p PeerID derived
 * from the Child's NeuronPublicKey (per 002 FR-006).
 */
function checkVREG12(
  p2p: readonly NeuronP2PExchangeEntry[],
  childPublicKey: NeuronPublicKey,
  errors: ValidationError[],
): void {
  const expectedPeerID = childPublicKey.peerId().toString();

  for (const svc of p2p) {
    if (svc.peerID && svc.peerID !== expectedPeerID) {
      errors.push({
        rule: 'V-REG-12',
        message:
          `neuron-p2p-exchange "${svc.name}" has peerID "${svc.peerID}" ` +
          `that does not match Child's PeerID. Expected: "${expectedPeerID}"`,
      });
    }
  }
}

File: impl/golang/internal/registry/validation.go
package registry

import (
	"fmt"
	"sync"

	"github.com/neuron-sdk/neuron-go-sdk/internal/keylib"
	"github.com/neuron-sdk/neuron-go-sdk/internal/payment"
)

// ValidationError represents a single validation rule violation during
// registration completeness checks.
type ValidationError struct {
	Rule    string // e.g., "V-REG-01"
	Message string
}

// Error implements the error interface.
func (e ValidationError) Error() string {
	return fmt.Sprintf("%s: %s", e.Rule, e.Message)
}

// AdmissionPolicy describes a registry's admission policy. The mechanism is
// platform-defined; this struct represents the policy's observable properties.
type AdmissionPolicy struct {
	PolicyType  string   `json:"policyType"`  // "permissionless" or "permissioned"
	TrustAnchor string   `json:"trustAnchor"` // e.g., "parent-did"
	Allowlist   []string `json:"allowlist"`   // Parent DIDs (for permissioned registries)
}

// AdmissionChecker defines the interface for admission policy implementations.
// Registries may use permissionless or allowlist-based policies.
type AdmissionChecker interface {
	IsAdmitted(childAddress string, parentDID string) (bool, error)
}

// PermissionlessPolicy always admits any registration attempt.
type PermissionlessPolicy struct{}

// IsAdmitted always returns true for a permissionless registry.
func (p PermissionlessPolicy) IsAdmitted(childAddress, parentDID string) (bool, error) {
	return true, nil
}

// AllowlistPolicy restricts registration to children whose parent DID is on
// the allowlist.
type AllowlistPolicy struct {
	mu        sync.RWMutex
	allowlist map[string]bool
}

// NewAllowlistPolicy creates an empty AllowlistPolicy.
func NewAllowlistPolicy() *AllowlistPolicy {
	return &AllowlistPolicy{
		allowlist: make(map[string]bool),
	}
}

// IsAdmitted returns true if the parentDID is on the allowlist.
func (p *AllowlistPolicy) IsAdmitted(childAddress, parentDID string) (bool, error) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if parentDID == "" {
		return false, AdmissionDenied{Detail: "parentDID is required for allowlist policy"}
	}
	if p.allowlist[parentDID] {
		return true, nil
	}
	return false, AllowlistRejection{ParentDID: parentDID}
}

// AddParentDID adds a parent DID to the allowlist.
func (p *AllowlistPolicy) AddParentDID(did string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.allowlist[did] = true
}

// RemoveParentDID removes a parent DID from the allowlist.
func (p *AllowlistPolicy) RemoveParentDID(did string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	delete(p.allowlist, did)
}

// Contains returns true if the given DID is on the allowlist.
func (p *AllowlistPolicy) Contains(did string) bool {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.allowlist[did]
}

// ValidateRegistrationCompleteness validates an AgentURI against all registration
// completeness rules. It collects ALL violations (never short-circuits) and
// returns (true, nil) when the registration is fully valid.
func ValidateRegistrationCompleteness(agentURI AgentURI, childPublicKey keylib.NeuronPublicKey) (bool, []ValidationError) {
	var errors []ValidationError

	// V-REG-01: Exactly 3 neuron-topic services.
	if len(agentURI.topicServices) != 3 {
		errors = append(errors, ValidationError{
			Rule:    "V-REG-01",
			Message: fmt.Sprintf("expected 3 neuron-topic services, got %d", len(agentURI.topicServices)),
		})
	}

	// V-REG-02: At least 1 neuron-p2p-exchange service.
	if len(agentURI.p2pServices) < 1 {
		errors = append(errors, ValidationError{
			Rule:    "V-REG-02",
			Message: "at least 1 neuron-p2p-exchange service is required",
		})
	}

	// V-REG-03: Each neuron-topic has MUST fields.
	for i, svc := range agentURI.topicServices {
		if svc.Type == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-03",
				Message: fmt.Sprintf("topic service[%d]: type must not be empty", i),
			})
		}
		if svc.Name == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-03",
				Message: fmt.Sprintf("topic service[%d]: name must not be empty", i),
			})
		}
		if svc.Version == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-03",
				Message: fmt.Sprintf("topic service[%d]: version must not be empty", i),
			})
		}
		if svc.Channel == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-03",
				Message: fmt.Sprintf("topic service[%d]: channel must not be empty", i),
			})
		}
		if svc.Transport == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-03",
				Message: fmt.Sprintf("topic service[%d]: transport must not be empty", i),
			})
		}
		if svc.Anchor == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-03",
				Message: fmt.Sprintf("topic service[%d]: anchor must not be empty", i),
			})
		}
		if svc.Config == nil {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-03",
				Message: fmt.Sprintf("topic service[%d]: config must not be nil", i),
			})
		}
	}

	// V-REG-04: Each neuron-p2p-exchange has MUST fields.
	for i, svc := range agentURI.p2pServices {
		if svc.Type == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-04",
				Message: fmt.Sprintf("p2p service[%d]: type must not be empty", i),
			})
		}
		if svc.Name == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-04",
				Message: fmt.Sprintf("p2p service[%d]: name must not be empty", i),
			})
		}
		if svc.Version == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-04",
				Message: fmt.Sprintf("p2p service[%d]: version must not be empty", i),
			})
		}
		if svc.PeerID == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-04",
				Message: fmt.Sprintf("p2p service[%d]: peerID must not be empty", i),
			})
		}
		if svc.Protocol == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-04",
				Message: fmt.Sprintf("p2p service[%d]: protocol must not be empty", i),
			})
		}
		if svc.TopicRef == "" {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-04",
				Message: fmt.Sprintf("p2p service[%d]: topicRef must not be empty", i),
			})
		}
	}

	// Build set of topic service names for cross-referencing.
	topicNames := make(map[string]bool)
	for _, svc := range agentURI.topicServices {
		if svc.Name != "" {
			topicNames[svc.Name] = true
		}
	}
	topicNameList := make([]string, 0, len(topicNames))
	for name := range topicNames {
		topicNameList = append(topicNameList, name)
	}

	// V-REG-05: Each p2p topicRef must match a neuron-topic service name.
	for i, svc := range agentURI.p2pServices {
		if svc.TopicRef != "" && !topicNames[svc.TopicRef] {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-05",
				Message: fmt.Sprintf("p2p service[%d]: topicRef %q not found in topic service names %v", i, svc.TopicRef, topicNameList),
			})
		}
	}

	// V-REG-06: DID (if present) matches childPublicKey.DIDKey().
	expectedDID := childPublicKey.DIDKey()
	for i, svc := range agentURI.didServices {
		if svc.Endpoint != expectedDID {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-06",
				Message: fmt.Sprintf("DID service[%d]: endpoint %q does not match expected %q", i, svc.Endpoint, expectedDID),
			})
		}
	}

	// V-REG-07: At most one DID service.
	if len(agentURI.didServices) > 1 {
		errors = append(errors, ValidationError{
			Rule:    "V-REG-07",
			Message: fmt.Sprintf("at most 1 DID service allowed, got %d", len(agentURI.didServices)),
		})
	}

	// V-REG-11: Each standard channel (stdIn, stdOut, stdErr) appears exactly once.
	channelCount := make(map[string]int)
	for _, svc := range agentURI.topicServices {
		if svc.Channel != "" {
			channelCount[svc.Channel]++
		}
	}
	for _, ch := range StandardChannels() {
		count := channelCount[ch]
		if count == 0 {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-11",
				Message: fmt.Sprintf("standard channel %q is missing", ch),
			})
		} else if count > 1 {
			errors = append(errors, ValidationError{
				Rule:    "V-REG-11",
				Message: fmt.Sprintf("standard channel %q appears %d times, expected exactly 1", ch, count),
			})
		}
	}

	// V-REG-12: Each p2p peerID matches childPublicKey.PeerID().String().
	expectedPeerID, peerErr := childPublicKey.PeerID()
	if peerErr == nil {
		for i, svc := range agentURI.p2pServices {
			if svc.PeerID != "" && svc.PeerID != expectedPeerID.String() {
				errors = append(errors, ValidationError{
					Rule:    "V-REG-12",
					Message: fmt.Sprintf("p2p service[%d]: peerID %q does not match expected %q", i, svc.PeerID, expectedPeerID.String()),
				})
			}
		}
	}

	// V-REG-13: Commerce service delivery cross-reference validation. FR-P01b.
	// Build set of p2p service names for cross-referencing.
	p2pNames := make(map[string]bool)
	for _, svc := range agentURI.p2pServices {
		if svc.Name != "" {
			p2pNames[svc.Name] = true
		}
	}

	for i, svc := range agentURI.commerceServices {
		switch svc.Delivery.Mode {
		case payment.DeliveryModeP2P:
			if svc.Delivery.ServiceRef != "" && !p2pNames[svc.Delivery.ServiceRef] {
				errors = append(errors, ValidationError{
					Rule:    "V-REG-13",
					Message: fmt.Sprintf("commerce service[%d] %q: delivery.serviceRef %q not found in p2p service names", i, svc.Name, svc.Delivery.ServiceRef),
				})
			}
		case payment.DeliveryModeTopic:
			if svc.Delivery.ChannelRef != "" && !topicNames[svc.Delivery.ChannelRef] {
				errors = append(errors, ValidationError{
					Rule:    "V-REG-13",
					Message: fmt.Sprintf("commerce service[%d] %q: delivery.channelRef %q not found in topic service names", i, svc.Name, svc.Delivery.ChannelRef),
				})
			}
		}
	}

	if len(errors) > 0 {
		return false, errors
	}
	return true, nil
}

// ValidateRoleBoundary enforces FR-R11 role-based operation constraints.
// Returns nil if the operation is allowed for the given role, or
// UnauthorizedOperation if the operation is forbidden.
func ValidateRoleBoundary(callerRole RegistryRole, operation string) error {
	switch callerRole {
	case RegistryAdmin:
		// Admin MUST NOT register, updateAgentURI, or transfer.
		switch operation {
		case "register", "updateAgentURI", "transfer":
			return UnauthorizedOperation{
				CallerRole: callerRole.String(),
				Operation:  operation,
			}
		}

	case RegisteredAgent:
		// Agent MUST NOT register for others.
		switch operation {
		case "register":
			return UnauthorizedOperation{
				CallerRole: callerRole.String(),
				Operation:  operation,
			}
		}

	case DelegatedOperator:
		// Operator MUST NOT register.
		switch operation {
		case "register":
			return UnauthorizedOperation{
				CallerRole: callerRole.String(),
				Operation:  operation,
			}
		}

	case ParentRole:
		// Parent MUST NOT register or updateAgentURI.
		switch operation {
		case "register", "updateAgentURI":
			return UnauthorizedOperation{
				CallerRole: callerRole.String(),
				Operation:  operation,
			}
		}

	default:
		return UnauthorizedOperation{
			CallerRole: callerRole.String(),
			Operation:  operation,
		}
	}

	return nil
}
